File: hashtag.c
/*Lauren Vogel    COP3514 Section2     Spring 2020              *
* The purpose of this program is to check an input to see if it *
* is a valid hashtag STARTING with the '#' character first, and *
* contains only letters, numbers, and underscores, no other     *
* special characters can be in the input. The program will print*
* whether the hashtag is valued or not before ending.           *
* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */

#include <stdio.h>

int main (){

        char ch;
        int valid = 1; /* Equal to 1 for true initialization */

        printf("Input: ");
        scanf("%c",&ch); /* Strictly to scan the first character only for '#' */
          if (ch != '#')
            valid = 0;
   
        ch = getchar();

        while(ch != '\n'){
          if (!((ch >= 'A' && ch <= 'Z') || (ch >= 'a' && ch <= 'z') || (ch >= '0' && ch <= '9') || (ch == '_')))
                valid = 0;
          ch = getchar();
        }
        /*Checks if valid is true or fasle */ 
        if (valid) (printf("Output: Is a valid hashtag\n"));
        else printf("Output: It is an Invalid hashtag\n");


  return 0;
}

File: hashtag.rb
=begin
  Lauren Vogel

 Practicing ruby basics by re-creating the
 hashtag assignment done in C.

 The purpose of this program is to check an input to see if it
 is a valid hashtag STARTING with the '#' character first, and
 contains only letters, numbers, and underscores, no other
 special characters can be in the input. The program will print
 whether the hashtag is valued or not before ending.

=end

#!/usr/bin/env ruby

puts "Check hashtag: "
hashtag = gets.chomp
#hashtag = hashtag.split("")

if hashtag[0] != '#'
  puts "Not a hashtag, missing '#'."
end

hashtag.each_char{ |c|
  #Regular expressionssss
  #\w is equivalent to [0-9a-zA-Z_]
  #\d is the same as [0-9]
  #\s matches white space (tabs, regular space, newline)
  #
  #There is also the negative form of these:
  #\W anything that’s not in [0-9a-zA-Z_]
  #\D anything that’s not a number
  #\S anything that’s not a space
  if c == "#" || c == "_" || c.match?(/\w/)
    next
  else
    return puts "Invalid hashtag containing illegal characters."
  end
}

puts hashtag + " is valid"
